pkg/networkmanager/firewall.jsx

```diff
@@ -0,0 +1,265 @@
+/* jshint esversion: 6 */
+
+/*
+ * This file is part of Cockpit.
+ *
+ * Copyright (C) 2018 Red Hat, Inc.
+ *
+ * Cockpit is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation; either version 2.1 of the License, or
+ * (at your option) any later version.
+ *
+ * Cockpit is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with Cockpit; If not, see <http://www.gnu.org/licenses/>.
+ */
+
+import cockpit from "cockpit";
+import React from "react";
+import firewall from "./firewall-client.es6";
+import { Listing, ListingRow } from "cockpit-components-listing.jsx";
+import { OnOffSwitch } from "cockpit-components-onoff.jsx";
+import { show_modal_dialog } from "cockpit-components-dialog.jsx";
+
+import "table.css";
+import "./networking.css";
+
+var _ = cockpit.gettext;
+
+require("page.css");
+
+function EmptyState(props) {
+    return (
+        <div className="curtains-ct blank-slate-pf">
+            {props.icon && <div className="blank-slate-pf-icon"><div className={props.icon}></div></div>}
+            <h1>{props.title}</h1>
+            {props.children}
+        </div>
+    );
+}
+
+function ServiceRow(props) {
+    if (!props.service.name)
+        return <ListingRow key={props.service.id}
+                           columns={["", "", ""]} />
+
+    var tcp = props.service.ports.filter(p => p.protocol.toUpperCase() == 'TCP');
+    var udp = props.service.ports.filter(p => p.protocol.toUpperCase() == 'UDP');
+
+    function onRemoveService(event) {
+        if (event.button !== 0)
+            return;
+
+        props.onRemoveService(props.service.id);
+        event.stopPropagation();
+    }
+
+    var columns = [
+        { name: props.service.name, header: true },
+        <div>
+            { tcp.map(p => p.port).join(', ') }
+        </div>,
+        <div>
+            { udp.map(p => p.port).join(', ') }
+        </div>,
+        <button className="btn btn-danger pficon pficon-delete"
+                onClick={onRemoveService}/>
+    ];
+
+    var tabs = [
+        { name: _("Details"), renderer: () => <p>{props.service.description}</p> }
+    ];
+
+    return <ListingRow key={props.service.id}
+                       rowId={props.service.id}
+                       columns={columns}
+                       tabRenderers={tabs}/>;
+}
+
+class AddServicesDialogBody extends React.Component {
+    constructor() {
+        super();
+
+        this.state = {
+            selected: new Set()
```

Comment:
Please initialize `services` here (to `null` so that your spinner still works). At least I prefer the state components to be spelled out here in full, for an easier mental model.